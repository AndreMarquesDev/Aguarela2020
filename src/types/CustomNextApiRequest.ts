import type { NextApiRequest } from 'next';
import type { FormPostRequestBody } from './FormPostRequestBody';

export type CustomNextApiRequest = {
    body: FormPostRequestBody;
} & NextApiRequest;
