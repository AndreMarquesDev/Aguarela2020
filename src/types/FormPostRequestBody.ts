import type { FormValues } from './FormValues';

export type FormPostRequestBody = {
    isPlaywright: boolean;
} & FormValues;
