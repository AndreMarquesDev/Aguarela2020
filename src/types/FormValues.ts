export type FormValues = {
    name: string;
    email: string;
    message: string;
    brand: string;
    subject: string;
};
