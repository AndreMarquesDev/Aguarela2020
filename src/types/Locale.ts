export enum Locale {
    Pt = 'pt',
    En = 'en',
}
