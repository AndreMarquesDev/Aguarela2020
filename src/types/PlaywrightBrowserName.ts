export enum PlaywrightBrowserName {
    Chromium = 'chromium',
    Firefox = 'firefox',
    Webkit = 'webkit',
}
