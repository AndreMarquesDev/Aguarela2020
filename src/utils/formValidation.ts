export enum FieldTypes {
    Name = 'name',
    Email = 'email',
    Textarea = 'message',
    Brand = 'brand',
    Subject = 'subject',
}
export type FormState = {
    name: string;
    email: string;
    message: string;
    brand: string;
    subject: string;
};

export const validateName = (value: string): boolean => {
    const nameRegex =
        // eslint-disable-next-line regexp/no-unused-capturing-group, regexp/prefer-plus-quantifier, regexp/prefer-character-class, regexp/no-super-linear-backtracking, regexp/no-useless-escape
        /s*([a-zA-ZáâàãéêèíîìóôòõúûùñçäöüßÁÂÀÃÉÊÈÍÎÌÓÔÒÕÚÛÙÑÇÄÖÜẞ]{1,}([\.,] |[-']| ))+[a-zA-ZáâàãéêèíîìóôòõúûùñçäöüßÁÂÀÃÉÊÈÍÎÌÓÔÒÕÚÛÙÑÇÄÖÜẞ]+\.?\s*$/;

    return nameRegex.test(value);
};

export const validateEmail = (value: string): boolean => {
    const emailRegex =
        // eslint-disable-next-line regexp/no-useless-escape, regexp/no-unused-capturing-group, regexp/prefer-d, regexp/use-ignore-case
        /(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    return emailRegex.test(value);
};
