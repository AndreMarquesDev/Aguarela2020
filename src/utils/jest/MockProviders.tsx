/* eslint-disable react/jsx-no-constructed-context-values */
import React, { ReactNode } from 'react';
import { NavLinksContext } from '../../components/context/NavLinksContext';
import { TextsContext } from '../../components/context/TextsContext';
import { Locale } from '../../types/Locale';
import { textsEn, textsPt } from '../texts';

interface MockProvidersProps {
    language?: Locale;
    isMenuOpen?: boolean;
    toggleMenu?: jest.Mock;
    setNavHeight?: jest.Mock;
    children: ReactNode;
}

export const MockProviders = ({
    language = Locale.Pt,
    isMenuOpen = false,
    toggleMenu = jest.fn(),
    setNavHeight = undefined,
    children,
}: MockProvidersProps): JSX.Element => {
    return (
        <TextsContext.Provider
            value={{
                texts: language === Locale.Pt ? textsPt : textsEn,
            }}
        >
            <NavLinksContext.Provider value={{ isMenuOpen, toggleMenu, setNavHeight }}>
                {children}
            </NavLinksContext.Provider>
        </TextsContext.Provider>
    );
};
