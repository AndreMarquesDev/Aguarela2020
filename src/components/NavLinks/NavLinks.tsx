import React, { useContext } from 'react';
import Link from 'next/link';
import classNames from 'classnames';
import { pagesMap } from '../../utils/pages';
import { TextsContext } from '../context/TextsContext';
import { LanguageButton } from '../LanguageButton/LanguageButton';
import { Locale } from '../../types/Locale';
import { TextsInterface } from '../../utils/texts';
import { phabletMediaQuery } from '../../styles/mediaQueries';

export interface NavLinksProps {
    currentRoute: string;
    language: Locale;
    isMobile: boolean;
    isMenuOpen: boolean;
}

export const NavLinks = ({
    currentRoute,
    language,
    isMobile,
    isMenuOpen,
}: NavLinksProps): JSX.Element => {
    const navLinks = pagesMap.filter(page => page !== 'homepage');
    const { texts } = useContext(TextsContext);

    const linksContainerStyles = classNames(
        isMobile && 'mobileLayout',
        isMenuOpen && isMobile && 'menuOpen'
    );
    const linkStyles = (link: string): string =>
        classNames('fontS', currentRoute.includes(link) && 'active');

    return (
        <ul className={linksContainerStyles}>
            {navLinks.map(link => (
                <li key={language + link}>
                    <Link href={`/${language}/${link}`} prefetch={false}>
                        {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
                        <a className={linkStyles(link)}>{texts[link as keyof TextsInterface]}</a>
                    </Link>
                </li>
            ))}

            <li className={classNames('languageButton fontXXS', isMobile && 'white fontXS')}>
                <LanguageButton />
            </li>

            <style jsx>
                {`
                    ul {
                        display: flex;

                        &.mobileLayout {
                            display: none;

                            li {
                                a {
                                    color: var(--white);
                                    margin-bottom: 25rem;
                                }
                            }
                        }

                        &.menuOpen {
                            width: 100%;
                            height: 100vh;
                            position: fixed;
                            display: flex;
                            flex-direction: column;
                            align-items: center;
                            justify-content: center;
                            top: 0;
                            left: 0;
                            z-index: 2;
                            background: var(--purple);
                        }
                    }

                    li {
                        margin: 0 25rem;

                        &:hover a,
                        .active {
                            color: var(--purple);
                        }

                        @media (${phabletMediaQuery}) {
                            margin: 0 calc(12.5rem / 2);
                        }

                        a {
                            display: inline-block;
                            text-transform: uppercase;
                            transition: color 0.15s linear;
                        }

                        &.languageButton {
                            align-self: center;
                            text-transform: uppercase;
                            margin-top: 5rem;
                            margin-left: 0;

                            &:hover {
                                color: var(--purple);
                            }

                            &.white {
                                color: var(--white);
                                margin-top: 0;
                            }

                            @media (${phabletMediaQuery}) {
                                margin: 5rem calc(12.5rem / 2) 0;
                            }
                        }
                    }
                `}
            </style>
        </ul>
    );
};
