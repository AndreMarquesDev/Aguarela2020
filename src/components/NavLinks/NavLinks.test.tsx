import type { RenderResult } from '@testing-library/react';
import type { NavLinksProps } from './NavLinks';
import { render, screen } from '@testing-library/react';
import React from 'react';
import { Locale } from '../../types/Locale';
import { MockProviders } from '../../utils/jest/MockProviders';
import { textsEn, textsPt } from '../../utils/texts';
import { NavLinks } from './NavLinks';
import '@testing-library/jest-dom';

const baseProps: NavLinksProps = {
    currentRoute: 'services',
    language: Locale.Pt,
    isMobile: false,
    isMenuOpen: false,
};

const renderComponent = (
    newProps?: Partial<NavLinksProps>,
    isMenuOpen = baseProps.isMenuOpen,
    language: Locale = Locale.Pt,
): RenderResult => {
    return render(
        <MockProviders isMenuOpen={isMenuOpen} language={language}>
            <NavLinks {...baseProps} {...newProps} />
        </MockProviders>,
    );
};

describe('<NavLinks />', () => {
    it('renders properly', () => {
        const { container } = renderComponent();

        const aboutLink = screen.getByText(textsPt.about);
        const projectsLink = screen.getByText(textsPt.projects);
        const servicesLink = screen.getByText(textsPt.services);
        const contactLink = screen.getByText(textsPt.contact);
        const languageButton = screen.getByText('pt');

        expect(aboutLink).toBeInTheDocument();
        expect(projectsLink).toBeInTheDocument();
        expect(servicesLink).toBeInTheDocument();
        expect(contactLink).toBeInTheDocument();
        expect(languageButton).toBeInTheDocument();

        expect(aboutLink).toHaveAttribute('href', `/${baseProps.language}/about`);
        expect(projectsLink).toHaveAttribute('href', `/${baseProps.language}/projects`);
        expect(servicesLink).toHaveAttribute('href', `/${baseProps.language}/services`);
        expect(contactLink).toHaveAttribute('href', `/${baseProps.language}/contact`);

        expect(container).toMatchSnapshot();
    });

    it('renders properly when language is en', () => {
        const { container } = renderComponent({ language: Locale.En }, false, Locale.En);

        const aboutLink = screen.getByText(textsEn.about);
        const projectsLink = screen.getByText(textsEn.projects);
        const servicesLink = screen.getByText(textsEn.services);
        const contactLink = screen.getByText(textsEn.contact);

        expect(aboutLink).toBeInTheDocument();
        expect(projectsLink).toBeInTheDocument();
        expect(servicesLink).toBeInTheDocument();
        expect(contactLink).toBeInTheDocument();

        expect(aboutLink).toHaveAttribute('href', '/en/about');
        expect(projectsLink).toHaveAttribute('href', '/en/projects');
        expect(servicesLink).toHaveAttribute('href', '/en/services');
        expect(contactLink).toHaveAttribute('href', '/en/contact');

        expect(container).toMatchSnapshot();
    });

    it('renders properly in mobile', () => {
        const { container } = renderComponent({ isMobile: true });

        expect(container).toMatchSnapshot();
    });

    it('renders properly when the menu is open', () => {
        const { container } = renderComponent({ isMenuOpen: true, isMobile: true }, true);

        expect(container).toMatchSnapshot();
    });
});
