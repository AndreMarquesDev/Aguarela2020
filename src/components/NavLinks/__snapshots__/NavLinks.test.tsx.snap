// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavLinks /> renders properly 1`] = `
<div>
  <ul
    class="jsx-2853090894 "
  >
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/pt/about"
      >
        Sobre
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/pt/projects"
      >
        Projetos
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS active"
        href="/pt/services"
      >
        Serviços
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/pt/contact"
      >
        Contacto
      </a>
    </li>
    <li
      class="jsx-2853090894 languageButton fontXXS"
    >
      <button
        class="jsx-2866958827"
        type="button"
      >
        pt
      </button>
    </li>
  </ul>
</div>
`;

exports[`<NavLinks /> renders properly in mobile 1`] = `
<div>
  <ul
    class="jsx-2853090894 mobileLayout"
  >
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/pt/about"
      >
        Sobre
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/pt/projects"
      >
        Projetos
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS active"
        href="/pt/services"
      >
        Serviços
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/pt/contact"
      >
        Contacto
      </a>
    </li>
    <li
      class="jsx-2853090894 languageButton fontXXS white fontXS"
    >
      <button
        class="jsx-2866958827"
        type="button"
      >
        pt
      </button>
    </li>
  </ul>
</div>
`;

exports[`<NavLinks /> renders properly when language is en 1`] = `
<div>
  <ul
    class="jsx-2853090894 "
  >
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/en/about"
      >
        About
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/en/projects"
      >
        Projects
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS active"
        href="/en/services"
      >
        Services
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/en/contact"
      >
        Contact
      </a>
    </li>
    <li
      class="jsx-2853090894 languageButton fontXXS"
    >
      <button
        class="jsx-2866958827"
        type="button"
      >
        pt
      </button>
    </li>
  </ul>
</div>
`;

exports[`<NavLinks /> renders properly when the menu is open 1`] = `
<div>
  <ul
    class="jsx-2853090894 mobileLayout menuOpen"
  >
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/pt/about"
      >
        Sobre
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/pt/projects"
      >
        Projetos
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS active"
        href="/pt/services"
      >
        Serviços
      </a>
    </li>
    <li
      class="jsx-2853090894"
    >
      <a
        class="jsx-2853090894 fontS"
        href="/pt/contact"
      >
        Contacto
      </a>
    </li>
    <li
      class="jsx-2853090894 languageButton fontXXS white fontXS"
    >
      <button
        class="jsx-2866958827"
        type="button"
      >
        pt
      </button>
    </li>
  </ul>
</div>
`;
